File: backend/src/main/java/com/example/backend/service/MovieService.java
package com.example.backend.service;

import com.example.backend.tmdb_client.client.MovieClient;
import com.example.backend.tmdb_client.dto.movies.BasicMovieDTO;
import com.example.backend.tmdb_client.dto.movies.MovieDetailsDTO;
import com.example.backend.tmdb_client.dto.movies.MovieTrailerDTO;
import com.example.backend.tmdb_client.response.movies.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class MovieService {

    private final MovieClient movieClient;

    public MoviesListResponse getPopularMovies() {
        return movieClient.getPopularMovies();
    }

    public MoviesListResponse getTrendingMovies() { return movieClient.getTrendingMovies(); }

    public NewReleaseMoviesListResponse getNewReleaseMovies() { return movieClient.getNewMovieReleases(); }

    public BasicMovieDTO getRandomMovie(){
        return movieClient.getRandomMovie().getResults().get(0);
    }

    public SearchMoviesListResponse getSearchMoviesResults(String query, int page) { return movieClient.getSearchMovieResults(query,page); }

    public DiscoverMoviesListResponse discoverMovies(Integer page, Integer genre, Integer yearFrom, Integer yearTo, Double rating, String sortBy) {
        return movieClient.discoverMovies(page, genre, yearFrom, yearTo, rating, sortBy);
    }

    public MovieTrailerDTO getMovieTrailer(Integer movieId) {

        MovieTrailerResponse response = movieClient.getMovieTrailer(movieId);

        if (response == null || response.getResults() == null) {
            return null;
        }

        return response.getResults().stream()
                .filter(video -> "YouTube".equals(video.getSite()))
                .filter(video -> "Trailer".equals(video.getType()))
                .filter(video -> Boolean.TRUE.equals(video.getOfficial()))
                .findFirst()

                .or(() ->
                        response.getResults().stream()
                                .filter(video -> "YouTube".equals(video.getSite()))
                                .filter(video -> "Trailer".equals(video.getType()))
                                .findFirst()
                )

                .or(() ->
                        response.getResults().stream()
                                .filter(video -> "YouTube".equals(video.getSite()))
                                .findFirst()
                )

                .orElse(null);
    }

    public MovieDetailsDTO getMovieDetails(Integer movieId) {

        MovieDetailsResponse response = movieClient.getMovieDetails(movieId);

        return MovieDetailsDTO.builder()

                .id(response.getId())
                .title(response.getTitle())
                .overview(response.getOverview())
                .runtime(response.getRuntime())
                .releaseDate(response.getReleaseDate())
                .rating(response.getVoteAverage())
                .posterPath(response.getPosterPath())
                .budget(response.getBudget())
                .revenue(response.getRevenue())
                .status(response.getStatus())
                .genres(response.getGenres())
                .productionCompanies(response.getProductionCompanies())
                .productionCountries(response.getProductionCountries())
                .spokenLanguages(response.getSpokenLanguages())
                .build();
    }
}

File: frontend/src/services/movieService.js
/*API*/
import api from "./api";

/*MOVIES*/
export async function getPopularMovies() {
  const response = await api.get("/movies/popular");
  return response.data.results;
}

export async function getTrendingMovies() {
  const response = await api.get("/movies/trending");
  return response.data.results;
}

export async function searchMovies(query, page = 1) {
    const { data } = await api.get("/movies/search", {
        params: {
            query,
            page
        }
    });
    return data;
}

export async function getNewMoviesReleases() {
  const response = await api.get("/movies/newReleases");
  return response.data.results;
}

export async function getRandomMovie() {
  const response = await api.get("/movies/random");
  return response.data;
}

export async function getMovieGenres() {
  const response = await api.get("/genres/movies");  
  return response.data;  
}

export async function discoverMovies(filters, page) {

    const { data } = await api.get("/movies/discover", {
        params: {
            page,
            genre: filters.genre || undefined,
            yearFrom: filters.yearFrom || undefined,
            yearTo: filters.yearTo || undefined,
            rating: filters.rating !== "all" ? filters.rating : undefined,
            sortBy: filters.sortBy || undefined,
        },
    });

    return data;
}

export async function getTrailer(mediaId, mediaType) {
  const { data } = await api.get(`/${mediaType}/${mediaId}/trailer`);
  return data;
}

export async function getMovieDetails(movieId) {
    const { data } = await api.get(`/movies/${movieId}/details`);
    return data;
}